wren-ai-service/src/pipelines/ask/historical_question.py

```diff
@@ -0,0 +1,90 @@
+import logging
+from typing import Dict, List, Optional
+
+from haystack import Document, Pipeline, component
+
+from src.core.document_store_provider import DocumentStoreProvider
+from src.core.llm_provider import LLMProvider
+from src.core.pipeline import BasicPipeline
+from src.utils import (
+    init_providers,
+    load_env_vars,
+)
+
+load_env_vars()
+logger = logging.getLogger("wren-ai-service")
+
+
+class HistoricalQuestion(BasicPipeline):
+    def __init__(
+        self, llm_provider: LLMProvider, store_provider: DocumentStoreProvider
+    ) -> None:
+        pipe = Pipeline()
+        pipe.add_component("embedder", llm_provider.get_text_embedder())
+        pipe.add_component(
+            "retriever",
+            store_provider.get_retriever(
+                document_store=store_provider.get_store(dataset_name="view_questions"),
+            ),
+        )
+        pipe.add_component("score_filter", ScoreFilter())
+        # todo: add a llm filter to filter out low scoring document
+        pipe.add_component("output_formatter", OutputFormatter())
+
+        pipe.connect("embedder.embedding", "retriever.query_embedding")
+        pipe.connect("retriever", "score_filter")
+        pipe.connect("score_filter", "output_formatter")
+
+        self._pipeline = pipe
+        super().__init__(self._pipeline)
+
+    def run(self, query: str):
+        logger.info("Try to extract historical question")
+        return self._pipeline.run(
+            {
+                "embedder": {"text": query},
+                "retriever": {"top_k": 1},
+                "score_filter": {"score": 0.8},
+            }
+        )
+
+
+@component
+class ScoreFilter:
+    @component.output_types(
+        documents=List[Document],
+    )
+    def run(self, documents: List[Document], score: float = 0):
+        return {
+            "documents": list(
+                filter(lambda document: document.score >= score, documents)
+            )
+        }
+
+
+@component
+class OutputFormatter:
+    @component.output_types(
+        documents=List[Optional[Dict]],
+    )
+    def run(self, documents: List[Document]):
+        list = []
+        logger.debug(f"historical_question_output_formatter: {documents}")
+
+        for doc in documents:
+            content = eval(doc.content)
```

Comment:
I am not sure the reason to use eval here

Reply:
after a discussion, we will use other function instead of python eval function to convert str to dict.